### DISCUSSION TITLE
Does solveOneNonlinearEquation really require the input being named "u"?

### DISCUSSION THREAD

**Discussion by bilderbuchi** [OpenModelica/OpenModelica#10879] [Q&A] [answered]
> `Modelica.Math.Nonlinear.solveOneNonlinearEquation` "determines the solution of one non-linear algebraic equation "y=f(u)" in one unknown "u" in a reliable way."

I read this such that I can feed it a function of one unknown, and it solves it for me. However, I've discovered that in OMEdit 1.22-dev-146, this fails if the _input_ variable of said function is not named `u` (e.g.  `x`), or if the _output_ variable is not named `y` (e.g. `out`).

Interpreting the docstring above strictly, the position could indeed be that only `u`/`y` is allowed. You would then have to use wrapper functions around your own functions to ensure that, but
1. this seems like a weird/arbitrary restriction (_especially considering Modelica's elegant features for partial function application_ ❤️) and
2. I read the source of `Modelica.Math.Nonlinear.solveOneNonlinearEquation`, and found nothing that indicates that it would need that hard-coded input/output names, as it just calls `f` with an argument.

MWE:
```modelica
model Solvenonlinear
parameter Real p1 = 0.1;

function phase
input Real u;
output Real y;
algorithm
  y := cos(u);
end phase;

function phase_fails
input Real x;
output Real y;
algorithm
  y := cos(x);
end phase_fails;

function phase_fails_out
input Real u;
output Real o;
algorithm
  o := cos(u);
end phase_fails_out;

Real u_res;
equation
//u_res = Modelica.Math.Nonlinear.solveOneNonlinearEquation(function phase(),0,3); // works

//u_res = Modelica.Math.Nonlinear.solveOneNonlinearEquation(function phase_fails(),0,3); // fails with
//  Type mismatch for positional argument 1 in Modelica.Math.Nonlinear.solveOneNonlinearEquation(f=Solvenonlinear.phase_fails). The argument has type:
//  Solvenonlinear.phase_fails<function>(Real) => Real
//expected type:
//  f<function>(#Real) => #Real

u_res = Modelica.Math.Nonlinear.solveOneNonlinearEquation(function phase_fails_out(),0,3); // fails, too

end Solvenonlinear;
```

Also, that error message is very unclear to me. I don't know what the relevance of that `#` is supposed to be?!

Is this a bug? If yes, I'll transfer this to the issues list.

**Comment by bilderbuchi** [comment]
Update: It turns out, that _in addition_, the output must be named `y`, too! (Edited original comment)

**Comment by perost** [comment, marked as answer]
[12.4.2](https://specification.modelica.org/master/functions.html#functional-input-arguments-to-functions) states that:
>the provided function must be function-compatible with the corresponding formal parameter of function type

[6.6](https://specification.modelica.org/master/interface-or-type-relationships.html#function-compatibility) states that one requirement for function compatibility is:
>All public input components of B have correspondingly named public input components of A in the same order and preceding any additional public input components of A.

So unfortunately it's a requirement that the inputs have the same names, and same for the outputs. As for the error message I agree that it's not very good. It should say why the types aren't compatible, but that's hard both to implement and present to the user in a good way.

The relevance of the `#` is that the type is boxed, which is an implementation detail that we shouldn't show users. I'll remove that since it's just confusing for users and not really that important for debugging anymore.

**Reply by perost** [reply]
#10881 removes the # from boxed types.

**Reply by bilderbuchi** [reply]
Aahhhh, I see now. Thanks for the clarification, that helps and is clear now!

Also, thanks for improving error messages. However, with #10881 merged, would not the above error message become/boil down to
```
The argument has type:
    my_f<function>(Real) => Real
  expected type:
    f<function>(Real) => Real
```
which is bound to be even more confusing? 😅  
As in: "it wants a function from Real to Real, and I give it a function from Real to Real, what is wrong here?"

**Reply by bilderbuchi** [reply]
potential improvement: include the function-compatibility that is the real problem here, if that is possible/visible to that part of the logging code:
```
The argument has type:
  my_f<function>(Real) => Real
expected type:
  f<function>(Real) => Real (function-compatible with Modelica.Math.Nonlinear.Interfaces.partialScalarFunction)
```
or
```
The argument has type:
  my_f<function>(Real) => Real
expected subtype of:
  f<Modelica.Math.Nonlinear.Interfaces.partialScalarFunction>(Real) => Real
```

**Reply by perost** [reply]
Yeah, as I said the error message isn't very good since it just prints the function signature without saying why they're not compatible. I don't think it's necessarily more confusing, just confusing in a different way.

The issue is that the type checking only returns whether a type is compatible or not, and not the reason for why it's not compatible. It's something I've wanted to improve for a long time, but never seem to find the time to actually do since it's quite a lot of work to change.
